equinox/nn/_attention.py

```diff
@@ -325,6 +401,42 @@ def __call__(
                 raise ValueError(
                     "process_heads must not change the shape of the heads."
                 )
+        state_length = None
+        if state is None:
+            causal_mask_offset = 0
+        else:
+            if self.kv_cache is None:
+                raise ValueError(
+                    "State was provided, but cannot use autoregressive decoding "
+                    "without specifying "
+                    "`MultiheadAttention(..., kv_cache=...)`. "
+                    "See `equinox.nn.StandardKVCache` for an example."
+                )
+
+            key_state, value_state, index, state = self.kv_cache(
+                state, key_heads, value_heads
+            )
+            _check_kv_shapes(key_state, value_state, key_heads, value_heads)
+            state_length = key_state.shape[0]
+
+            causal_mask_offset = index - kv_seq_length
+            key_heads = key_state
+            value_heads = value_state
+            kv_seq_length = state_length
+
+        if mask == "causal":
+            query_indices = jnp.arange(query_seq_length)[:, None]
+            kv_indices = jnp.arange(kv_seq_length)[None, :]
+            mask = kv_indices <= query_indices + causal_mask_offset
+        if state is not None and state_length is not None:
```

Comment:
I think `state is not None` should imply `state_length is not None`, right?

If you just need to communicate this fact to a static type checker then this can be done using an assert:
```python
if state is not None:
    assert state_length is not None
    ...
```